# Nguồn
1. https://phamdinhkhanh.github.io/2019/08/10/PytorchTurtorial1.html#21-autograd-t%E1%BB%B1-%C4%91%E1%BB%99ng-t%C3%ADnh-%C4%91%E1%BA%A1o-h%C3%A0m
2.https://minhng.info/ai/pytorch-co-ban.html

# Nội dung kiến thức
1. Giới thiệu về Pytorch
2. Tensor
3. Các toán tử trên Tensor
4. Broadcast
5. Variable <br/>
    5.1 Khởi tạo Variable <br/>
    5.2 Autograd <br/>

# 1. Giới thiệu về Pytorch

Chào mọi người, đây là bài viết đầu tiên của năm 2021 và cũng là bài trở lại sau nửa năm bận rộn và tẩu hỏa của mình :D. Mình trở lại để viết tiếp những chuỗi bài về AI để chia sẻ kiến thức cũng như tự củng cố cho bản thân. Và hôm nay chúng ta tìm hiểu về Pytorch.

Pytorch là 1 framework được xây dựng trên python cung cấp nền tảng tính toán trong lĩnh vực Deep learning, cùng với Tensorflow, Pytorch đang là framework được cộng đồng Deep learning ưa chuộng nhất hiện nay.

Tensorflow thì được ưa chuộng để build lên các sản phẩm thực tế, deploy mô hình, áp dụng vào các hệ thống lớn. Còn Pytorch thì có vẻ được ưa chuộng trong nghiên cứu. Về cá nhân mình khi tìm hiểu cả 2 thì mình có hứng thú với pytorch hơn vì cảm thấy tính rõ ràng trong nó. Phong cách code pytorch tương đồng python. Và nó sở hữu khả năng tính toán GPU mạnh mẽ, mang lại hiệu năng tính toán cao.

Bây giờ chúng ta hãy cùng xem xét pipeline cơ bản của một dự án Pytorch. Hình dưới đây mô tả một quy trình công việc điển hình cùng các modules quan trọng được liên kết ở mỗi bước.
![1](./images/1.png)

Có thể thấy các modules cơ bản của pytorch như là: **torch.utils.data, torch.nn, torch.optim, torch.utils, torch.autograd...**

Chúng ta sẽ cùng tìm hiểu chi tiết thông qua các ví dụ ở các bài sau.

Về hướng dẫn cài đặt, các bạn có thể xem trên trang chủ của pytorch.

# 2. Tensor
Tensor là kiểu dữ liệu chính trong Pytorch, có thể hình dung Tensor là khối ma trận nhiều chiều. Việc khởi tạo và truy xuất tương tự numpy array.

In [1]:
# Khởi tạo tensor ngẫu nhiên có kích thước 3 hàng 4 cột
import torch
x = torch.Tensor(3,4)
print(x)

tensor([[-4.3005e-29,  1.2083e-38,  1.1590e-26,  2.3036e-39],
        [-4.2990e-35,  1.2171e-38,  6.4620e+16, -2.8977e-27],
        [ 4.8123e-38, -2.7642e-27,  1.6132e+16, -4.7328e-26]])


In [2]:
# khởi tạo tensor phân phối đều
x = torch.rand(3,4)
print(x)

tensor([[0.0515, 0.2511, 0.0280, 0.4123],
        [0.5357, 0.6541, 0.4171, 0.1434],
        [0.9501, 0.2531, 0.2801, 0.4306]])


In [3]:
# Khởi tạo tensor nhiều chiều
x = torch.Tensor([[2,5], [4,3], [1,5]])
print(x)
print(x.size()) # Kiểm tra kích cỡ Tensor

tensor([[2., 5.],
        [4., 3.],
        [1., 5.]])
torch.Size([3, 2])


In [4]:
# khởi tạo ma trận 0 với data type là long
x = torch.zeros(3, 4, dtype=torch.long)
print(x)

tensor([[0, 0, 0, 0],
        [0, 0, 0, 0],
        [0, 0, 0, 0]])


In [6]:
# Khởi tạo từ kiểu dữ liệu numpy
import numpy as np
x = np.array([1,2,3])
y = torch.from_numpy(x)

print(x)

print(y)

y += 1
print(x)
print(y)

[1 2 3]
tensor([1, 2, 3])
[2 3 4]
tensor([2, 3, 4])


Khi chuyển đổi x thành Tensor y. Mà tensor y cộng thêm 1 đơn vị thì x cũng tự động thay đổi theo. Chúng ta cũng có thể chuyển đổi ngược tensor về numpy như sau:

In [7]:
# Chuyển đổi tensor sang numpy
z = y.numpy()
print(z)
print(type(z))

[2 3 4]
<class 'numpy.ndarray'>


Có nhiều cách khởi tạo nữa, khá giống với Tensorflow và numpy.

Tính toán trên GPU nhanh gấp nhiều lần CPU. Tensor hỗ trợ việc tính toán trên GPU, mặc định Tensor tạo theo cách thông thường sẽ nằm trên CPU. Để đem Tensor lên GPU tính toán, chúng ta dùng phương thức  **.cuda()**

In [8]:
x = torch.Tensor([1,2,3])
y = torch.Tensor([4,5,6])
if torch.cuda.is_available(): # kiểm tra xem máy tính hỗ trợ CUDA ko
    x = x.cuda()
    y = y.cuda()
    print(x + y)
else:
    print('Cuda is not available!')

tensor([5., 7., 9.], device='cuda:0')


Xem thêm: https://pytorch.org/docs/0.3.0/tensors.html#torch-tensor

# 3. Các toán tử trên Tensor

* Resize: *view()*
* Phép cộng: *torch.add(a,b) = a + b*
* Phép trừ: *a.sub(b) = a - b*
* Phép nhân tương ứng từng phần tử: *torch.mul(a,b) = a * b*
* Phép chia tương ứng từng phần tử: *torch.div(a,b) = a / b*
* Mean: *a.mean()*
* Độ lệch tiêu chuẩn (std): *a.std()*
* Chuyển tensor 1 phần tử sang numeric python: *item()*

In [9]:
x = torch.randn(3,3)
print(x)

y = torch.randn(3,3)
print(y)

tensor([[ 1.6835,  0.0084,  0.3737],
        [ 0.1589,  0.7473,  0.4028],
        [-0.5508,  0.2677,  0.5509]])
tensor([[ 1.9175, -0.6483, -0.6887],
        [-0.4578,  0.0642,  0.8489],
        [-1.4211,  0.6726, -0.7297]])


In [10]:
print("x+y:", x+y)
print("x-y:", x-y)
print("x*y:", x*y)
print("x/y:", x/y)
print("mean of x:", x.mean())
print("std of y:", y.std())

x+y: tensor([[ 3.6010, -0.6399, -0.3150],
        [-0.2989,  0.8116,  1.2517],
        [-1.9719,  0.9404, -0.1789]])
x-y: tensor([[-0.2341,  0.6566,  1.0624],
        [ 0.6167,  0.6831, -0.4461],
        [ 0.8703, -0.4049,  1.2806]])
x*y: tensor([[ 3.2281, -0.0054, -0.2574],
        [-0.0728,  0.0480,  0.3419],
        [ 0.7827,  0.1801, -0.4020]])
x/y: tensor([[ 0.8779, -0.0129, -0.5427],
        [-0.3471, 11.6326,  0.4745],
        [ 0.3876,  0.3980, -0.7549]])
mean of x: tensor(0.4047)
std of y: tensor(1.0301)


In [12]:
# reshape x
z = x.view(9)
print(z)
print(z.size())

tensor([ 1.6835,  0.0084,  0.3737,  0.1589,  0.7473,  0.4028, -0.5508,  0.2677,
         0.5509])
torch.Size([9])


In [13]:
# Chuyển tensor 1 phaafnf tử sang numeric python
x = torch.tensor([1.5])
i = x.item()
print(i, type(i))

1.5 <class 'float'>


# 4. Broadcast
Broadcast cho phép ta thực hiện các toán tử giữa 2 tensor của pytorch/mảng numpy không cùng số chiều. Quy tắc Broadcast như sau:
* Thêm chiều vào phía trước tensor có chiều ngắn hơn cho đến khi 2 tensor bằng nhau.
* Số phần tử của mỗi chiều của 2 tensor phải:
    - bằng nhau, hoặc
    - một tensor có một phần tử, giá trị của phần tử được nhân bản thành k phần tử (bằng số phần tử của tensor còn lại ở chiều đang xét)

In [14]:
x = torch.rand(2,1,3)
y = torch.ones(3,1)
print(x)
print(y)
print(x+y)

tensor([[[0.4644, 0.0326, 0.3631]],

        [[0.5650, 0.8192, 0.5750]]])
tensor([[1.],
        [1.],
        [1.]])
tensor([[[1.4644, 1.0326, 1.3631],
         [1.4644, 1.0326, 1.3631],
         [1.4644, 1.0326, 1.3631]],

        [[1.5650, 1.8192, 1.5750],
         [1.5650, 1.8192, 1.5750],
         [1.5650, 1.8192, 1.5750]]])


Giải thích:
* Kích thước của x: 2x1x3
* Kích thước của y: 3x1
* Broadcast:
    - Thêm chiều vào phía trước của y cho đến khi số chiều bằng x --> y là 1x3x1
    - Xét chiều thứ 0: Chiều 0 của y là 1, nhỏ hơn x (là 2) --> kích thước y: 2x3x1
    - Xét chiều thứ 1: Chiều 1 của y là 3, lớn hơn x (là 1) --> Kích thước x: 2x3x3
    - Xét chiều thứ 2: Chiều 2 của y là 1, nhỏ hơn x (là 3) --> Kích thước y: 2x3x3
* Kết quả kích thước tensor đầu ra: 2x3x3


Nếu vi phạm quy tắc broadcast sẽ báo lỗi.

(Khi thêm chiều y: 1x3x1x1, x là: 5x2x4x1 --> Vi phạm ở chiều 1 của x và y - không bằng nhau số chiều mà 1 trong 2 cũng không phải là tensor 1 phần tử ở chiều 1 đó)

In [15]:
x=torch.FloatTensor(5,2,4,1)
y=torch.FloatTensor(3,1,1)
(x+y).size()

RuntimeError: The size of tensor a (2) must match the size of tensor b (3) at non-singleton dimension 1

# 5. Variable

Variable là class bao bọc (wrapper) Tensor cho phép thực hiện tính toán đạo hàm. Variable lưu trữ **data** (tensor) và **grad** (gradient)
![img](./images/2.png)

## 5.1 Khởi tạo Variable

In [16]:
from torch.autograd import Variable
x = Variable(torch.ones(2,2), requires_grad=True)
print(x)

tensor([[1., 1.],
        [1., 1.]], requires_grad=True)


Mỗi một tensor torch sẽ có 1 thuộc tính là **.requires_grad**, nếu bạn set thuộc tính này về True, các toán tử triển khai trên tensor sẽ được theo dõi. Khi kết thúc quá trình lan truyền thuận (hoặc quá trình tính toán output) bạn có thể gọi **.backward()** và mọi tính toán gradient sẽ được tự động thực hiện dựa trên lịch sử đã được lưu lại. Các gradient cho tensor này sẽ được tích lũy và xem tại thuộc tính **.grad**.

Để dừng theo dõi một tensor chúng ta gọi vào hàm **.detach()**. Khi đó các hoạt động trên tensor sẽ không còn được lưu vết nữa.

Ngoài ra để ngăn tensor lưu lại lịch sử (và sử dụng memory), chúng ta cũng có thể bao quanh code block triển khai tensor với hàm **with torch.no_grad()**: nó rất hữu ích trong trường hợp đánh giá model bởi vì khi thuộc tính *requires_grad = True* thì model sẽ có thể được cập nhật tham số. Nhưng **quá trình đánh giá model sẽ không cần cập nhật tham số** nên chúng ta không cần áp dụng gradient lên chúng. Đơn giản là set *requires_grad = False*.

2 class **Tensor** và **Function** cùng tương tác và xây dựng một đồ thị chu trình mà đồ thị này mã hóa lại toàn bộ lịch sử tính toán. Mỗi một tensor đều có một thuộc tính **grad_fn** trích dẫn đến một **Function** đã tạo ra Tensor (Trừ trường hợp tensor được tạo ra bởi người dùng được set thuộc tính grad_fn là None).

Nếu muốn tính toán đạo hàm chúng ta gọi hàm **.backward()** của Tensor. Nếu Tensor là một scalar sẽ không cần xác định bất kỳ đối số **gradient** nào cho .backward(). Tuy nhiên, khi tensor có nhiều hơn 1 phần tử cần xác định đối số **gradient** là một tensor có cùng kích thước. Đối số này sẽ quy định tốc độ thay đổi theo **gradient** tại mỗi chiều là bao nhiêu.

## 5.2 Autograd

Minh họa cách tính Gradient tự động trong Pytorch bằng ví dụ sau: Cho đồ thị tính toán như hình. Đầu vào **[x,y,z] = [5,3,7]**. Tất cả các trọng số w đều được khởi tạo là **w=0.5**. Tính kết quả (biến result) khi ta cho [x,y,z] qua mạng (forward) và gradient của mỗi trọng số w (backward).
![img](./images/3.png)

In [19]:
from torch.autograd import Variable

xy = Variable(torch.FloatTensor([5, 3]), requires_grad=True) # size: 2
z = Variable(torch.FloatTensor([7]), requires_grad=True)    # size: 1
w12 = Variable(torch.FloatTensor([0.5, 0.5]), requires_grad=True)   # size: 2
w3 = Variable(torch.FloatTensor([0.5]), requires_grad=True) # size: 1
w4 = Variable(torch.FloatTensor([0.5]), requires_grad=True) # size: 1
w5 = Variable(torch.FloatTensor([0.5]), requires_grad=True) # size: 1

In [21]:
k = xy*w12
print(k)

tensor([2.5000, 1.5000], grad_fn=<MulBackward0>)


In [22]:
o1 = k[0] * k[1]
print(o1)

tensor(3.7500, grad_fn=<MulBackward0>)


In [23]:
o2 = o1*w4+z*w3
print(o2)

tensor([5.3750], grad_fn=<AddBackward0>)


In [24]:
result = o2*w5
print(result)

tensor([2.6875], grad_fn=<MulBackward0>)


In [25]:
result.backward()
print('w12.grad', w12.grad)     # 1.8750, 1.8750
print('w3.grad', w3.grad)       # 3.5000
print('w4.grad', w4.grad)       # 1.8750
print('w5.grad', w5.grad)       # 5.3750

w12.grad tensor([1.8750, 1.8750])
w3.grad tensor([3.5000])
w4.grad tensor([1.8750])
w5.grad tensor([5.3750])


Giải thích:

**Forward:**
- o1 = x * w1 * y * w2 = 5 * 0.5 * 3 * 0.5 = 3.75
- o2 = o1 * w4 + z * w3 = 3.75 * 0.5 + 7 * 0.5 = 5.375
- result = o2 * w5 = 5.375 * 0.5 = 2.6875

**Backward:**
- Đạo hàm result theo w5 = (o2 * w5)' = o2 = 5.375
- Đạo hàm result theo o2 = w5 = 0.5
- Đạo hàm result theo w3 = (o2 * w5)' = ((o1 * w4 + z * w3) * w5)' = z * w5 = 3.5

Tương tự cho w4,w1 và w2